addon/-private/state-machine/-base.ts

```diff
@@ -47,24 +56,126 @@ export default abstract class StateMachine extends EmberObject {
   }
 
   addStep(this: StateMachine, name: string) {
-    if (this.stepTransitions.includes(name)) {
+    if (this.stepsToAdd.includes(name)) {
       return;
     }
 
-    // Set the first step, if it hasn't been yet
-    if (!get(this, 'firstStep')) {
-      set(this, 'firstStep', name);
+    this.stepsToAdd.push(name);
+    return bind(
+      this,
+      scheduleOnce,
+      'afterRender',
+      this,
+      this.flushAdditionQueue
+    )();
+  }
+
+  removeStep(this: StateMachine, name: string) {
+    if (this.stepsToRemove.includes(name)) {
+      return;
     }
 
-    this.stepTransitions.pushObject(name);
+    this.stepsToRemove.push(name);
+    return bind(this, scheduleOnce, 'render', this, this.flushRemoveQueue)();
```

Comment:
This seems suspect too, I don't really like returning a runloop-bound function that itself it scheduling another function to run later on.

Reply:
This is for unit tests entirely. All of the integration tests pass, but I get a runloop error associated with the linear and circular state machine unit tests.

Reply:
Hmm, okay. I'll dig into this more then and see if I can work out what's happening.